Source/WebCore/dom/ViewTransition.cpp

```diff
@@ -250,14 +255,37 @@ void ViewTransition::captureOldState()
         // FIXME: Fill in the rest of CapturedElement.
         CapturedElement capture;
         auto transitionName = element->computedStyle()->viewTransitionName();
-        m_namedElements.set(transitionName->name, capture);
+        m_namedElements.add(transitionName->name, capture);
     }
+    return { };
 }
 
 // https://drafts.csswg.org/css-view-transitions/#capture-new-state-algorithm
-void ViewTransition::captureNewState()
+ExceptionOr<void> ViewTransition::captureNewState()
 {
-    // FIXME: Implement this.
+    if (!m_document)
+        return { };
+    ListHashSet<AtomString> usedTransitionNames;
+    Ref document = *m_document;
+    CheckVisibilityOptions visibilityOptions { .contentVisibilityAuto = true };
+    // FIXME: Loop should probably use flat tree.
+    for (Ref element : descendantsOfType<Element>(document)) {
+        if (!element->checkVisibility(visibilityOptions))
+            continue;
+        ASSERT(element->computedStyle());
+        auto transitionName = element->computedStyle()->viewTransitionName();
+        if (!transitionName)
+            continue;
+        auto& name = transitionName->name;
+        if (usedTransitionNames.contains(name))
+            return Exception { ExceptionCode::InvalidStateError, makeString("Multiple elements found with view-transition-name: "_s, name) };
+        usedTransitionNames.add(name);
+        if (!m_namedElements.contains(name))
+            m_namedElements.add(name, CapturedElement { });
+        m_namedElements.find(name)->newElement = element.ptr();
```

Comment:
Do we actually need `usedTransitionNames`? Isn't the presence of `newElement` in `m_namedElements[name]` the same?